### PR TITLE
Add samba backend

### LINKED ISSUE #308: [Feature Request] Samba support
Hi,

I've got a small PR for adding support for SAMBA based on the excellent [go-smb2](https://pkg.go.dev/github.com/hirochachacha/go-smb2) library.

Currently I'm facing build issues with plg_image_light on Linux, for some reason it's complaining about missing references to libglib2.0. Help here would be appreciated - it builds when commenting out the necessary code.

Cheers,
Ben

### PR DESCRIPTION
**Pull request by ThreeFx**
Closes #308.

To be resolved:

- [x] Testing with samba servers
- [x] `gosmb2` seems to allow only `\` separators. Maybe we have to string-replace `/` to `\`?

### CODE REVIEW

**Comment by mickael-kerjean** [comment]
Brilliant, thank you so much for this!!!!!!!!!

**Comment by ThreeFx** [comment]
I think that this is feasible like this. I haven't tested:

* multiple concurrent accesses (I only have access to one account on the share I tested)
* If `toSambaPath` is correct for all inputs. I suspect that it should be the case but I'm not 100% sure.

@Glorfischi mentioned he wanted to take a peek at the code sometime.

**Comment by mickael-kerjean** [comment]
> multiple concurrent accesses (I only have access to one account on the share I tested)

I don't have access to a samba share to give it a go. Do you know of some publicly accessible samba share I could try to give this a go? Also, that's definitely not a blocker as far as merging the code. 

> If toSambaPath is correct for all inputs. I suspect that it should be the case but I'm not 100% sure.

Taking from my experience working out with some other protocols like FTP and SFTP, different server implementation tend to handle things slightly differently from one another which sometimes brake things in a subtle way. Ultimately it's hard to test every single implementation of a protocol and the way I've been doing this so far is to release things and see what people report over time, slowly fixing around the edge cases.

> @Glorfischi mentioned he wanted to take a peek at the code sometime.

I'm not sure to follow, what do you mean?

> Samba permission errors are thrown as "Not found", only the error popup in the bottom left indicates that the user has no permission to do that action. Is there a way to signal to the application that this should be a 403?

yes, there's a bunch of predefined errors you can use from [here](https://github.com/mickael-kerjean/filestash/blob/master/server/common/error.go#L14-L30). On the plus side when you return any of those, it's getting translated in any of [those languages](https://github.com/mickael-kerjean/filestash/tree/master/client/locales)

File: server/plugin/plg_backend_samba/index.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	share *smb2.Share
+}
+
+func (smb Samba) Init(params map[string]string, app *App) (IBackend, error) {
+	c := SambaCache.Get(params)
+	if c != nil {
+		return c.(*Samba), nil
+	}
+
+	conn, err := net.DialTimeout("tcp", params["host"]+":445", 10 * time.Second)
```
Comment (Glorfischi) [inline server/plugin/plg_backend_samba/index.go:40]: The port shouldn't be hardcoded.

File: server/plugin/plg_backend_samba/index.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+func (smb Samba) Touch(path string) error {
+	path = toSambaPath(path)
+	return smb.Touch(path)
+}
+
+func toSambaPath(path string) string {
+	path = strings.TrimLeft(path, `/`)
+	return strings.Replace(path, `/`, `\`, -1) // replace all path separators
```
Comment (Glorfischi) [inline server/plugin/plg_backend_samba/index.go:162]: I'm a fan of [ReplaceAll](https://golang.org/pkg/strings/#example_ReplaceAll). I think it's clearer.

**Review by Glorfischi — COMMENTED** [review]
LGTM in general. 

I think we should definitely do some more testing. 

Also run go fmt before merging, it's not formatted correctly.

**Comment by ThreeFx** [comment]
> > multiple concurrent accesses (I only have access to one account on the share I tested)
> 
> I don't have access to a samba share to give it a go. Do you know of some publicly accessible samba share I could try to give this a go? Also, that's definitely not a blocker as far as merging the code.

Sadly I don't, and I also can't find any with a quick google search. The SMB share I tested this with is from my university, which has some enterpris-ey IBM/Windows thing. It works well as far as I can tell.

> > If toSambaPath is correct for all inputs. I suspect that it should be the case but I'm not 100% sure.
> 
> Taking from my experience working out with some other protocols like FTP and SFTP, different server implementation tend to handle things slightly differently from one another which sometimes brake things in a subtle way. Ultimately it's hard to test every single implementation of a protocol and the way I've been doing this so far is to release things and see what people report over time, slowly fixing around the edge cases.

Sounds good to me - we'll upstream any fixes we do for our case.

> > Samba permission errors are thrown as "Not found", only the error popup in the bottom left indicates that the user has no permission to do that action. Is there a way to signal to the application that this should be a 403?
> 
> yes, there's a bunch of predefined errors you can use from [here](https://github.com/mickael-kerjean/filestash/blob/master/server/common/error.go#L14-L30). On the plus side when you return any of those, it's getting translated in any of [those languages](https://github.com/mickael-kerjean/filestash/tree/master/client/locales)

Thanks, I switched to `ErrPermissionDenied` for that case. Other than that I'd say we upstream new errors as we go.

Should be ready to merge from my end.

**Comment by ThreeFx** [comment]
Drafted again because I was experiencing strange errors. Turns out somebody was playing around with permissions when I was testing. _sigh_. I will undraft once I've verified that everything works correctly.

**Comment by ThreeFx** [comment]
I tested once the people were done playing around and it seems to work. I can

* create files
* edit/save files
* rename files
* upload files
* delete files
* create directories
* rename directories
* delete directories

Renaming directories is a bit weird since the client seems to open the old directory, which of course does not exist anymore, a similar thing happens with files. I suspect this is not an issue with the SAMBA implementation.

Undrafting for the hopefully final time ;)

**Review by mickael-kerjean — APPROVED** [review]
LGTM
